aptos-move/aptos-vm-types/src/change_set.rs

```diff
@@ -358,6 +367,40 @@ impl VMChangeSet {
         Ok(())
     }
 
+    fn squash_additional_resource_writes(
+        write_set: &mut HashMap<StateKey, (WriteOp, Option<MoveTypeLayout>)>,
+        additional_write_set: HashMap<StateKey, (WriteOp, Option<MoveTypeLayout>)>,
+    ) -> anyhow::Result<(), VMStatus> {
+        for (key, additional_entry) in additional_write_set.into_iter() {
+            match write_set.entry(key) {
+                Occupied(mut entry) => {
+                    // Squash entry and addtional entries if type layouts match
+                    let (additional_write_op, additional_type_layout) = additional_entry;
+                    let (write_op, type_layout) = entry.get_mut();
+                    if *type_layout != additional_type_layout {
+                        return Err(VMStatus::error(
+                            StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                            err_msg("Cannot squash two writes with different type layouts."),
+                        ));
+                    }
+                    let noop = !WriteOp::squash(write_op, additional_write_op).map_err(|e| {
```

Comment:
There is squash_writes_pair, maybe we can reuse that/refactor a bit?

Reply:
There used to be squash_writes_pair previously. Now, I had to expand that for this function, because here `WriteOp::Squash` takes `write_op` as input and in the else statement, I need to remove `entry`. That wasn't possible with squash_writes_pair.

Reply:
Why? this part of the code seems identical to me

Reply:
Macro:
```
($write_entry:ident, $additional_write:ident) => {
        // Squashing creation and deletion is a no-op. In that case, we
        // have to remove the old write op from the write set.
        let noop = !WriteOp::squash($write_entry.get_mut(), $additional_write).map_err(|e| {
            VMStatus::error(
                StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
                err_msg(format!("Error while squashing two write ops: {}.", e)),
            )
        })?;
        if noop {
            $write_entry.remove();
        }
    };
```
and here
```
let noop = !WriteOp::squash(write_op, additional_write_op).map_err(|e| {
    VMStatus::error(
        StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
        err_msg(format!("Error while squashing two write ops: {}.", e)),
    )
})?;
if noop {
    entry.remove();
}
```
seems like adapting the code slightly would solve this